File: Atcoder_#314/javascript/c_rotate_colored_subsequence.js
function main(input) {
    const lines = input.trim().split('\n');
    const [n, m] = lines[0].split(' ').map(Number);
    let s = lines[1].split('');
    const c = lines[2].split(' ').map(Number);

    // 各色に対応するインデックスのリストを作成
    const positionsByColor = Array.from({ length: m + 1 }, () => []);

    for (let i = 0; i < n; i++) {
        positionsByColor[c[i]].push(i);
    }

    // 各色ごとに処理
    for (let color = 1; color <= m; color++) {
        const positions = positionsByColor[color];
        if (positions.length === 0) continue;

        // 右に1つシフトする
        const lastChar = s[positions[positions.length - 1]];
        for (let i = positions.length - 1; i > 0; i--) {
            s[positions[i]] = s[positions[i - 1]];
        }
        s[positions[0]] = lastChar;
    }

    // 結果を出力
    console.log(s.join(''));
}

// 標準入力を読み込み、main関数を呼び出し
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin', 'utf8');
main(input);

File: Atcoder_#314/ruby/c_rotate_colored_subsequence.rb
# 入力を受け取る
n, m = gets.split.map(&:to_i)
s = gets.chomp
c = gets.split.map(&:to_i)

# 各色に対応するインデックスのリストを作成
positions_by_color = Array.new(m + 1) { [] }

n.times do |i|
  positions_by_color[c[i]] << i
end

# 結果を格納する文字列の配列を初期化
result = s.chars

# 各色ごとに処理
(1..m).each do |color|
  positions = positions_by_color[color]
  next if positions.empty?

  # 右に1つシフトする
  last_char = result[positions[-1]]
  (positions.size - 1).downto(1) do |i|
    result[positions[i]] = result[positions[i - 1]]
  end
  result[positions[0]] = last_char
end

# 結果を出力
puts result.join
